program/src/processor.rs

```diff
@@ -210,9 +232,124 @@ fn process_distribute_rewards(
-    _accounts: &[AccountInfo],
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
 ) -> ProgramResult {
+    let accounts_iter = &mut accounts.iter();
+
+    let holder_rewards_pool_info = next_account_info(accounts_iter)?;
+    let holder_rewards_info = next_account_info(accounts_iter)?;
+    let token_account_info = next_account_info(accounts_iter)?;
+    let mint_info = next_account_info(accounts_iter)?;
+    let _system_program = next_account_info(accounts_iter)?;
+
+    let token_supply = {
+        let mint_data = mint_info.try_borrow_data()?;
+        let mint = StateWithExtensions::<Mint>::unpack(&mint_data)?;
+        mint.base.supply
+    };
+
+    let token_account_balance = {
+        let token_account_data = token_account_info.try_borrow_data()?;
+        let token_account = StateWithExtensions::<Account>::unpack(&token_account_data)?;
+
+        // Ensure the provided token account is for the mint.
+        if !token_account.base.mint.eq(mint_info.key) {
+            return Err(PaladinRewardsError::TokenAccountMintMismatch.into());
+        }
+
+        token_account.base.amount
+    };
+
+    let last_seen_total_rewards = {
+        // Ensure the holder rewards pool is owned by the Paladin Rewards
+        // program.
+        if !holder_rewards_pool_info.owner.eq(program_id) {
+            return Err(ProgramError::InvalidAccountOwner);
+        }
+
+        // Ensure the provided holder rewards pool address is the correct
+        // address derived from the mint.
+        if !holder_rewards_pool_info
+            .key
+            .eq(&get_holder_rewards_pool_address(mint_info.key))
+        {
+            return Err(PaladinRewardsError::IncorrectHolderRewardsPoolAddress.into());
+        }
+
+        let holder_rewards_pool_data = holder_rewards_pool_info.try_borrow_data()?;
+        let holder_rewards_pool_state =
+            bytemuck::try_from_bytes::<HolderRewardsPool>(&holder_rewards_pool_data)
+                .map_err(|_| ProgramError::InvalidAccountData)?;
+
+        holder_rewards_pool_state.total_rewards
+    };
+
+    // Calculate unharvested rewards for the token account.
+    //
+    // Since the holder rewards account is being initialized, the
+    // `unharvested_rewards` is calculated from the _available_ rewards in the
+    // pool, ie. `pool.lamports - rent_exempt_minimum`.
+    //
+    // If the program used total rewards for this calculation, new holders
+    // would be able to claim rewards that were already distributed to other
+    // holders.
+    //
+    // If the program used zero rewards for this calculation, new holders
+    // would not be able to claim rewards until the next distribution, which
+    // could result in some lamports left unclaimable in the pool.
+    let unharvested_rewards = {
+        let rent = <Rent as Sysvar>::get()?;
+        let rent_exempt_lamports = rent.minimum_balance(std::mem::size_of::<HolderRewardsPool>());
+        let active_rewards = holder_rewards_pool_info
```

Comment:
nit: maybe call this `available_rewards`? I'm not sure what `active` means

Reply:
"active rewards" I plucked from the program's original design document, but you're right "available" makes more sense.